mongodb-nosqli.yaml

```diff
@@ -0,0 +1,109 @@
+rules:
+- id: mongodb-nosqli
+  message: Detected non-constant data passed into a NoSQL query using the 'where'
+    evaluation operator. If this data can be controlled by an external user, this
+    is a NoSQL injection. Ensure data passed to the NoSQL query is not user controllable,
+    or properly sanitize the data. Ideally, avoid using the 'where' operator at all
+    and instead use the helper methods provided by com.mongodb.client.model.Filters
+    with comparative operators such as eq, ne, lt, gt, etc.
+  languages:
+  - java
+  severity: WARNING
+  patterns:
+  - pattern-either:
+    - pattern: (BasicDBObject $QUERY).put("$where", $INPUT);
+    - pattern: '(HashMap<String, String> $MAP).put("$where", $INPUT);
+
+        ...
+
+        (BasicDBObject $QUERY).putAll($MAP);
+
+        '
+    - pattern: (BasicDBObject $QUERY).append("$where", $INPUT);
+    - pattern: new BasicDBObject("$where", $INPUT);
+    - pattern: '(HashMap<String, String> $MAP).put("$where", $INPUT);
+
+        ...
+
+        new BasicDBObject($MAP);
+
+        '
+    - pattern: '(HashMap<String, String> $MAP).put("$where", $INPUT);
+
+        ...
+
+        String json = new JSONObject($MAP).toString();
+
+        (BasicDBObject $QUERY).parse((String $JSON));
+
+        '
+    - pattern: BasicDBObjectBuilder.start().add("$where", $INPUT);
+    - pattern: BasicDBObjectBuilder.start().append("$where", $INPUT);
+    - pattern: BasicDBObjectBuilder.start("$where", $INPUT);
+    - pattern: '(HashMap<String, String> $MAP).put("$where", $INPUT);
+
+        ...
+
+        BasicDBObjectBuilder.start($MAP);
+
+        '
+  - pattern-not: (BasicDBObject $QUERY).put("$where", "...");
```

Comment:
All your `pattern-not` serve the purpose of filtering out cases where `$INPUT` is a literal string: you can replace them all with the following block and achieve the same result in a more concise way!
```
  - metavariable-pattern:
      metavariable: $INPUT
      patterns:
        - pattern: |
            ...
        - pattern-not: |
            "..."
```

Reply:
excellent, that makes things more maintainable

Reply:
Fixed in 67842403b5064740657a5810ea9576bc15f1f84c